core/shared/src/main/scala/fs2/Stream.scala

```diff
@@ -774,6 +774,101 @@ final class Stream[+F[_], +O] private (private val free: FreeC[Algebra[Nothing,
     go(s, this).stream
   }
 
+  /**
+    * Like [[Stream#evalMap]], but will evaluate effects in parallel, emitting the results
+    * downstream in the same order as the input stream. The number of concurrent effects
+    * is limited by the `maxConcurrent` attribute and by size of chunk, whichever is lower.
+    *
+    * For example chunks of size 1 will always get evaluated synchronously, even when `maxConcurrent`
+    * will be set to Int.MaxValue.
+    *
+    * Similarly, if the chunk has size of 100, but `maxConcurrent` is set to 1, then effects will be evaluated
+    * sequentially.
+    *
+    *
+    * If any of concurrently running computations fails, the remainder of the computations iimmediately
+    * cancelled, and ocne all computations are cancelled, this will fail.
+    *
+    * @param maxConcurrent  Maximum number of concurrent effects.
+    * @param f              A function to transform `O` to `O2` evaluating effect `F`
+    */
+  def evalMapPar[F2[x] >: F[x]: Concurrent, O2](maxConcurrent: Int)(
+      f: O => F2[O2]): Stream[F2, O2] = {
+    def go(remains: Vector[O]): F2[Vector[O2]] = {
+      val (head, tail) = remains.splitAt(maxConcurrent)
+      if (head.isEmpty) Applicative[F2].pure(Vector.empty)
+      else {
+        head.traverse(o => Concurrent[F2].start(f(o))).flatMap { fibers =>
+          fibers.traverse(_.join).attempt.flatMap {
+            case Right(v) =>
+              go(tail).map(v ++ _)
+            case Left(err) =>
+              fibers.traverse_(_.cancel).attempt >> Sync[F2].raiseError(err)
+          }
+        }
+      }
+    }
+
+    this.chunks.flatMap { chunk =>
+      Stream.evalUnChunk(go(chunk.toVector).map(Chunk.vector))
+    }
+  }
+
+  /**
+    * Like [[Stream#evalMap]], but will evaluate effects in parallel, emitting the results
+    * downstream. The number of concurrent effects is limited by the `maxConcurrent` attribute
+    * and by size of chunk, whichever is lower.
+    *
+    * If any of concurrently running computations fails, the remainder of the computations iimmediately
+    * cancelled, and ocne all computations are cancelled, this will fail.
+    *
+    * See [[Stream#evalMapPar]] if retaining the original order of the stream is required.
+    *
+    */
+  def evalMapParUnordered[F2[x] >: F[x]: Concurrent, O2](maxConcurrent: Int)(
```

Comment:
Seems like using this method on a stream with singleton chunks results in the same behavior as `evalMap`, which I find unintuitive. That is, I don't think chunk size should impact number of concurrent effects running.

Reply:
@SystemFw current implementation runs essentially whole lot of orchestration for processing the single element. This is not necessarily, hence for example you don't need any resource management, interruption etc. for `F[B]`

Just to be on same page. Are you talking about Unordered variant or both ?

Reply:
When maxConcurrent == chunkSize, doesn't this implementation wait until each of the tasks finish before taking the next chunk? This is different behavior than the old version which attempted to make sure there were *always* maxConcurrent tasks running, regardless of chunk structure.

Reply:
@mpilquist Ah, I see what you mean. Sorry this took me that long to realize. Ok then, lets wait a while and I gonna improve this :-)